### ISSUE TITLE
useTrail works differently if modified in an effect as opposed to directly in canary.808.17

### ISSUE THREAD

**Issue by HQ92** [react-spring/react-spring#942]
## 🐛 Bug Report

My useTrail implementation in a custom hook powers the animating of content in once they enter the viewport on a site and used to work fine in an earlier beta version of react-spring 9. Ever since upgrading to canary.808.17 the hook started animating all the springs in at once instead of staggering them like before.

I can't see a reason why and have tried a few things. I've managed to get the new useSpring, useSprings and useTransition hooks working with the new api but can't see any reason why useTrail has changed the way it works. I've made a simplified repo which I think demonstrates the issue.

The spring animates correctly if modified directly outside of an effect, but doesn't if updated through an effect. The reason I want to do it inside an effect is so that i can wrap the trail spring into a function which passes in a prop that tells the spring when it is in the viewport in order to trigger the spring to start animating.

## To Reproduce

Try animating useTrail within a react useEffect hook

## Expected behavior

Expect it to work in the same way as updating the trail would work outside of the effect.

## Link to repro (highly encouraged)

https://codesandbox.io/s/awesome-glitter-vxd61?fontsize=14&hidenavigation=1&theme=dark

The repo also seems to reset the spring correctly if the spring is only updated from outside the effect directly, but if you run the effect version then even the direct update to the spring doesn't work as expected.

## Environment

- `react-spring` v9.0.0-canary.808.17.55c5691
- `react` v16.8.6

**Comment by HQ92** [comment]
The closest way I could find of replicating useTrail to work with useEffect is based on dbismut's implementation at https://github.com/react-spring/react-spring/issues/739#issuecomment-529994764. But I'm not sure if that is more or less efficient than the way that useTrail used to work.

``` javascript
  const [animateIn, setAnimateInSpring] = useSprings(amount, i => ({
    active: 0,
    default: true,
    onChange: () => {
      setAnimateInSpring(i => {
        if (i !== 0) {
          return ({
            active: animateInSpring[i - 1].active.get(),
          });
        }
      });
    },
  }));

  React.useEffect(() => {
    setAnimateInSpring((i, spring) => {
      if (i !== 0) return;
      return ({
        active: triggerAnimateIn? 1 : 0,
      });
    });
  }, [triggerAnimateIn]);
```

**Comment by aleclarson** [comment]
Fixed in `9.0.0-rc.2` (#985) without any code changes required. 🎉 

https://codesandbox.io/s/react-spring-942-6opnk

**Comment by HQ92** [comment]
Thanks! Great work on this library, it really has made it so much easier to enhance user experience through motion with its intuitive implementation.
